src/analyzer/protocol/tcp/TCP.cc

```diff
@@ -1307,47 +1305,132 @@ int TCP_Analyzer::ParseTCPOptions(const struct tcphdr* tcp,
 
 		else if ( options + 1 >= opt_end )
 			// We've run off the end, no room for the length.
-			return -1;
+			break;
 
 		else
 			opt_len = options[1];
 
 		if ( opt_len == 0 )
-			return -1;	// trashed length field
+			break;	// trashed length field
 
 		if ( options + opt_len > opt_end )
 			// No room for rest of option.
-			return -1;
-
-		if ( (*proc)(opt, opt_len, options, analyzer, is_orig, cookie) == -1 )
-			return -1;
+			break;
 
+		opts.emplace_back(options);
 		options += opt_len;
 
 		if ( opt == TCPOPT_EOL )
 			// All done - could flag if more junk left over ....
 			break;
 		}
 
-	return 0;
-	}
-
-int TCP_Analyzer::TCPOptionEvent(unsigned int opt,
-					unsigned int optlen,
-					const u_char* /* option */,
-					TCP_Analyzer* analyzer,
-					bool is_orig, void* cookie)
-	{
 	if ( tcp_option )
+		for ( const auto& o : opts )
+			{
+			auto kind = o[0];
+			auto length = kind < 2 ? 1 : o[1];
+			ConnectionEventFast(tcp_option, {
+				BuildConnVal(),
+				val_mgr->GetBool(is_orig),
+				val_mgr->GetCount(kind),
+				val_mgr->GetCount(length),
+				});
+			}
+
+	if ( tcp_options )
 		{
-		analyzer->ConnectionEventFast(tcp_option, {
-			analyzer->BuildConnVal(),
+		auto option_list = new VectorVal(BifType::Vector::TCP::OptionList);
+
+		for ( const auto& o : opts )
+			{
+			auto kind = o[0];
+			auto length = kind < 2 ? 1 : o[1];
+			auto option_record = new RecordVal(BifType::Record::TCP::Option);
+			option_list->Assign(option_list->Size(), option_record);
+			option_record->Assign(0, val_mgr->GetCount(kind));
+			option_record->Assign(1, val_mgr->GetCount(length));
+
+			auto handled_option = false;
+
+			switch ( kind ) {
+			case 2:
+				// MSS
+				if ( length == 4 )
+					{
+					auto mss = ntohs(*reinterpret_cast<const uint16_t*>(o + 2));
+					option_record->Assign(3, val_mgr->GetCount(mss));
+					handled_option = true;
+					}
```

Comment:
Would it perhaps make sense to raise weird events in cases where the length does not match what we expect? I know we do still report them - but we don't make it that easy to show that something potentially weird is up.

Reply:
Yeah, seems that could be helpful; pushed a new commit to add those weirds.